Add Function.experimental_get_tracing_count fixing #37323

PlusPlusUltra opened the pull request:
Issue #37323 requested a public API to know how many times a function was traced.

This PR turns the previously private `_get_tracing_count` method into a public API, documents it properly, and adds unit tests to check if it is working correctly.

The public API has been named `experimental_get_tracing_count` with the `experimental_` prefix to better control the adoption of the API.

PlusPlusUltra committed:
Add Function.experimental_get_tracing_count

At the moment, the Function class has an internal only method to
retrieve the number of times a function was traced. Some users
want to access this value externally.

This commit turns the internal method into a public API (with the
experimental_ prefix to control the API's adoption), and documents
it properly.

PlusPlusUltra committed:
Add test for experimental_get_tracing_count

PlusPlusUltra commented:
@googlebot I signed it!

mdanatg commented on tensorflow/python/eager/def_function_test.py:
```diff
@@ -938,6 +938,17 @@ def f(x):
     self.assertIn('Tracing is expensive', logs.output[0])
+  
+  def test_experimental_get_tracing_count(self):
+    @tf.function
```
You need to use def_function.function here - def_function is just the internal namespace for tf.function.

PlusPlusUltra replied:
Thanks for the feedback, I am addressing both of your corrections and will push the changes soon. I just wonder, do you prefer me to amend my existing commits or to create new commits with my fixes?
I'd like to be compliant with the overall tensorflow style.

mdanatg replied:
New commits should be fine, they make it easier to browse the PR history. Thanks!

mdanatg commented on tensorflow/python/eager/def_function_test.py:
```diff
@@ -938,6 +938,17 @@ def f(x):
     self.assertLen(logs.output, 1)
     self.assertIn('Tracing is expensive', logs.output[0])
+  
+  def test_experimental_get_tracing_count(self):
```
Since the tracing counter is global, it would be useful to test it on a method as well. Specifically:

```
class TestClass:
  @tf.function
  def method(...):
    ...

obj1 = TestClass()
# cause some retraces on obj1.method...
# verify counters for obj1.method

obj2 = TestClass()
# cause some retraces on obj2.method...
# verify counters for obj2.method
# then verify counters for obj1.method again
```

mdanatg reviewed (commented):
Looks good, just needs a bit more testing.

PlusPlusUltra committed:
More tests for experimental_get_tracing_count

The function experimental_get_tracing_count was further tested.
A method was defined in a class, and experimental_get_tracing_count
was used to count the number of tracings of said method in two different
instances of said class. As expected, the counts in the two instances
were independent from each other.
A minor fix was made in the previous test.

mdanatg commented on tensorflow/python/eager/def_function.py:
````diff
@@ -773,7 +773,41 @@ def _decorate(self, decorator):
+    
+    For example if this function is defined:
+
+    ```@tf.function
````
We have a mechanism that ensures these snippets are tested. It uses [doctest](https://docs.python.org/3/library/doctest.html). Let's write the code snippets using that syntax as well.

Here's an example:

https://github.com/tensorflow/tensorflow/blob/70e52bd49af344df268c25077a00b0b758982f16/tensorflow/python/eager/def_function.py#L1315

Also, feel free to point to the tf.function documentation to make sure the information is not duplicated.

PlusPlusUltra replied:
Thanks again! I will push the new changes as soon as I can. The topic is discussed in this guide https://www.tensorflow.org/guide/function. I suppose it is fine to point to this guide instead?

mdanatg replied:
Yep that sounds good.

mdanatg commented on tensorflow/python/eager/def_function_test.py:
```diff
@@ -938,6 +938,35 @@ def f(x):
     self.assertLen(logs.output, 1)
     self.assertIn('Tracing is expensive', logs.output[0])
+  
+  def test_experimental_get_tracing_count(self):
```
Let's break this into two tests:

```
test_experimental_get_tracing_count_function
test_experimental_get_tracing_count_method
```

mdanatg reviewed (commented):
Just a couple more formatting edits, otherwise looks good.

PlusPlusUltra committed:
Add doctest tests

The code snippets in the comments were edited so that they will be
tested by doctest. A link to a guide was provided that explains better
some details about the retracing of functions.
The tests were split. Now the tests for when
experimental_get_tracing_count is used on a function, and the one for
when it is used in a method are separated.

PlusPlusUltra commented:
@mdanatg should I do something about the CI checks failing? I see that on the main branch some commits fail some of them as well. So maybe I just forked the branch when those were already failing?

mdanatg commented on tensorflow/python/eager/def_function_test.py:
```diff
@@ -938,6 +938,36 @@ def f(x):
+    def double(a):
+      return a + a
+    
+    double(tf.constant(1))
```
This needs to be `constant_op.constant` instead of `tf.constant` - that's what causes the CI failures.

PlusPlusUltra replied:
oops, I am sorry, I will fix it immediately

PlusPlusUltra replied:
Ok, I made a new commit, I should have changed the old commit?

mdanatg replied:
No need - new commits are actually better.

PlusPlusUltra committed:
Fix an error

An error was corrected. Now the internal namespace for tf.constant
is used, as it should.

mdanatg commented on tensorflow/python/eager/def_function_test.py:
```diff
@@ -938,6 +938,36 @@ def f(x):
+    
+    double(constant_op.constant(1))
+    double(constant_op.constant(2))
+    assertAllEqual(double.experimental_get_tracing_count(), 1)
```
Sorry, missed this - it needs to be `self.assertAllEqual`.

mdanatg commented on tensorflow/python/eager/def_function.py:
```diff
@@ -773,7 +773,35 @@ def _decorate(self, decorator):
     self._function_spec = function_lib.FunctionSpec.from_function_and_signature(
         self._python_function, self.input_signature)
 
-  def _get_tracing_count(self):
```
It looks like there are a few user of this method, unfortunately (one is in `tensorflow/python/distribute/client/client.py`). There are two options:
 * try a global search for `_get_tracing_count` and update all uses 
 * keep both methods (and just have one call the other), along with a TODO to remove the private one

mdanatg reviewed (changes requested):
A couple more things that the tests flagged.

PlusPlusUltra committed:
Keep the private _get_tracing_count method
The private function _get_tracing_count was used by other functions
so renaming it caused errors.
Now the private function simply returns experimental_get_tracing_count.
A TODO was added, to remove the private method once it is not useful
anymore.
An alternative was to just updated all the uses of the private method,
so that the public function is called instead, but it's probably
better to keep the private method until the public method has the
experimental label.

PlusPlusUltra committed:
Fix typo in previous commit

PlusPlusUltra committed:
Merge branch 'master' into PlusPlusUltra-add-trace-count

rthadur commented:
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.
